packages/hint-doctype/src/hint.ts

```diff
@@ -0,0 +1,125 @@
+        docs: {
+            category: Category.interoperability,
+            description: `Hint to validate if the doctype is correct`
+        },
+        id: 'doctype',
+        schema: [],
+        scope: HintScope.any
+    }
+
+    public constructor(context: HintContext) {
+
+        const doctypeRegexFactory = (flags?: string) => {
+            return new RegExp(/(<!doctype\s+(html)\s*?)(>)(.+)?/, flags ? flags : 'g');
+        };
+
+        const checkDoctypeIsValid = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<boolean> => {
+            debug(`Checking if the doctype is valid.`);
+
+            const matched = content.match(doctypeRegexFactory('gi'));
+
+            if (!matched || matched.length < 1) {
+                await context.report(resource, element, `The file does not contain a valid doctype tag`);
+
+                return false;
+            }
+
+            return true;
+        };
+
+        const checkDoctypeFirstLine = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<void> => {
+            debug(`Checking if the doctype is in the first line.`);
+
+            const firstLine = content.split(/\r|\n/)[0];
+            const matched = firstLine.match(doctypeRegexFactory('gi'));
+
+
+            if (!matched || matched.length < 1) {
+                await context.report(resource, element, `The first line does not contain a valid doctype tag.`);
+
+                return;
+            }
+
+            if (matched) {
+                const cleaned = matched[0].trim().split('');
+
+                if (cleaned[cleaned.length-1] !== '>') {
+                    await context.report(resource, element, `There is additional information on the line with the doctype tag`);
+
+                    return;
+                }
+            }
+        };
+
+        const checkDoctypeLowercase = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<void> => {
+            debug(`Checking that the doctype is in lowercase`);
+
+            const matched = content.match(doctypeRegexFactory());
+
+            if (!matched) {
+                await context.report(resource, element, `The doctype should be in lowercase`);
+
+                return;
+            }
+        };
+
+        const checkNoDuplicateDoctype = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<void> => {
+            debug(`Checking that there is only one doctype tag in the document`);
+
+            const matched = content.match(doctypeRegexFactory('gi'));
+
+            if (matched && matched.length > 1) {
+                await context.report(resource, element, `There is more than one doctype tag in the document`);
+
+                return;
+            }
+        };
+
+        const onFetchEndHTML = async (fetchEnd: FetchEnd): Promise<void> => {
+            const { resource, element, response } = fetchEnd;
```

Comment:
The `element` here will be the one that triggered the request, not the `doctype` in this `resource`. We'll need to either search for the actual node for the `doctype` to pass it on to `context.report` (which I'm not sure is possible in all our connectors) or explicitly pass a `ProblemLocation` to `context.report` based on where the issue was found.

Reply:
> (which I'm not sure is possible in all our connectors) or explicitly pass a `ProblemLocation` to `context.report` based on where the issue was found.

Probably calculating the `ProblemLocation` is going to be the more robust solution assuming the doctype is going to be always near the top...

Reply:
Do you want to configure that the ProblemLocation always on the first line?
Something like:
```
export type ProblemLocation = {
    column: 0;
    line: 0;
};
```
Or do you prefer to search for the doctype tag and if this is in the 10th line calculate that the problem location is in the 10th line?

Reply:
> Or do you prefer to search for the doctype tag and if this is in the 10th line calculate that the problem location is in the 10th line?

I'd go with this unless we don't find a doctype at all, in which case the first line is fine. This should make it clearer to developers what needs to change, especially in cases where there's multiple doctypes or content after the doctype.

Reply:
Yes, thanks.
Only to be sure, this is how you expect the report for now?
```
await context.report(resource, null, 'Resource has no content.', undefined, defaultProblemLocation);
```
We didn't find how to do it without the null and the undefined.

Reply:
Yes, we probably need to rework the available signatures for `report` a bit, but that's correct given the way things stand today.

I filed https://github.com/webhintio/rfcs/issues/47 to track making this better in the future.

Reply:
Updated with this